Yoast/wordpress-seo
# Use React Modal for modals

Trying to recap what this PR does:

- each modal is based on a configuration object passed via PHP
- I guess in the future this object can be passed via JS
- it is possible to have multiple modals in the same page: the `window.yoastModalConfig` global variable is actually an array of objects so there will be just one `yoastModalConfig` in a page, containing the arguments for all the modals in a page
- modals have a standardized "frame" made of some default elements: 
  - open button 
  - heading
  - close "X" button 
  - close button at the bottom
- each element can be controlled from the PHP args and displayed/omitted when necessary
- buttons can have optional class names
- a11y: modals have a `role="dialog"` and a required `aria-label` attribute; other a11y requirements are taken care by `react-modal`


To add a new modal:

- build a component to be used as the modal content
- import it to `js/src/components/modals/index.js` and add it to the `modals` object. See https://reactjs.org/docs/jsx-in-depth.html#choosing-the-type-at-runtime
- so far, the modal content component needs localized strings to be passed from PHP
- thus, any component used as modal content needs a class/method or function to get and enqueue the translations
- use `IntlProvider` when needed
- initialize the modal from PHP `Yoast_Modal::add( $args );` passing args as documented in `Yoast_modal::get_defaults()`
- to pass to `react-modal` the mount point to hook to, you can use any CSS query selector, for example: `'mountHook' => '.inside .wpseo-metabox-buy-premium'` or `'mountHook' => '#something > .first-level-child'`, or anything else that targets an element in the page: the element **content** will be replaced by the modal "Open" button.


Note 1:
In order to print the localized strings when they're needed, I'm using the action hook `admin_footer` for `wp_localize_script()`. Please notice this is not exactly the indended usage of `wp_localize_script()`, but seems to work well.

Note 2:
The second modal `ModalTestContent` is meant just for testing / demo purposes and should be removed before merge

In order to test:
- switch to this branch
- switch yoast-component to `stories/af/5781-react-yoast-modal` and yarn-link it
- on wordpress-seo, run `composer du` and `yarn install`. Rebuild the assets, just in case.
- edit a post
- in the metabox, click Add Keyword to see the "Get Yoast SEO Premium now!" modal and click on the "Second modal" button to see a second modal (used just for demo purposes)

<img width="984" alt="screen shot 2018-03-23 at 09 27 03" src="https://user-images.githubusercontent.com/1682452/37819121-b60c707e-2e7c-11e8-9b5c-96b4738ac0cf.png">

See also the modal content fully translated:

<img width="682" alt="screen shot 2018-03-28 at 09 33 32" src="https://user-images.githubusercontent.com/1682452/38019009-cc9d5da0-3276-11e8-972c-28b8e9691003.png">


Fixes #5781
